### PR TITLE
Brendan dev

### LINKED ISSUE #2: create-plan-tmp-hdf
Given an existing HEC-RAS Plan HDF file (`.p**.hdf`), create a new temp Plan HDF file (`.p**.tmp.hdf`). For example:
```
$ ./ras_remodeler.py create-plan-tmp-hdf Muncie.p04.hdf Muncie.p04.tmp.hdf
```
Resulting temp Plan HDF file (e.g. `Muncie.p04.tmp.hdf`) should be identical to the same temp Plan HDF file produced by HEC-RAS when the associated unsteady flow plan is run via the RAS GUI. These temp Plan HDF files are necessary for running RAS in a Linux environment. It looks like the primary (only?) difference between the `.p**.hdf` and `.p**.tmp.hdf` files is that the latter are missing a "Results" section.

Use the Muncie example model for testing.

### LINKED ISSUE #3: set-plan-hdf-hydrograph
Given an existing HEC-RAS plan HDF file, overwrite an existing hydrograph in the HDF file with a hydrograph from a HEC-DSS file output from HEC-HMS. For example:
```
$ ./ras_remodeler.py set-hdf-hydrograph Muncie.p04.tmp.hdf "River: White  Reach: Muncie  RS: 15696.24" ./hms-output.dss://internal/dss/path/to/some/hydrograph
```

Where `"River: White  Reach: Muncie  RS: 15696.24"` is the name of an existing hydrograph in `Muncie.p04.tmp.hdf`.

* For now, assume that all HDF attributes of the hydrograph will remain the same. Only the values in the array will be replaced.
* Use the Muncie example model for testing.

### LINKED ISSUE #4: Dockerfile
Create an initial `Dockerfile` for `wat-ras-remodeler`. Probably nearly identical to the Dockerfile from `wat-hydrograph-stats-py`: https://github.com/water-tech-repos/wat-hydrograph-stats-py/blob/main/Dockerfile

### PR DESCRIPTION
**Pull request by barne856**

### CODE REVIEW

File: tests/test_hdf.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            assert group1 == group2
+            n_rows += 1
+        assert n_rows == 3
+    with h5py.File(temp_file2, 'r') as file:
+        n_rows = 0
+        for group1, group2 in zip(file.keys(), ["Event Conditions", "Geometry", "Plan Data", "Results"]):
+            assert group1 == group2
+            n_rows += 1
+        assert n_rows == 4
+    os.remove(temp_file1)
```
Comment (thwllms) [inline tests/test_hdf.py:30]: Advise using `pytest` features enable a proper "teardown" where `temp_file1` and `temp_file2` are deleted after the test even if the test fails.

File: tests/test_hdf.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    """Test copy attributes in hdf files"""
+    src_file = "tests/data/Muncie.p04.hdf"
+    temp_file = get_file(ext=".hdf")
+    copy_hdf(src_file, temp_file, ["Results"])
+    with h5py.File(src_file, 'r') as src, h5py.File(temp_file, 'r+') as temp:
+        dataset1 = src["/Event Conditions/Unsteady/Boundary Conditions/Flow Hydrographs/River: White  Reach: Muncie  RS: 15696.24"]
+        dataset2 = temp["/Event Conditions/Unsteady/Boundary Conditions/Flow Hydrographs/River: White  Reach: Muncie  RS: 15696.24"]
+        copy_attrib(dataset1, dataset2, "Interval", "Mins")
+        assert dataset2.attrs["Interval"].decode() == "Mins"
+    os.remove(temp_file)
```
Comment (thwllms) [inline tests/test_hdf.py:58]: See above re: removing temporary test files

File: tests/test_hdf.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    temp_file = get_file(ext=".hdf")
+    copy_hdf(src_file, temp_file, ["Results"])
+    timeseries = read_dss_timeseries(
+        "tests/data/hydrograph.dss:/REGULAR/TIMESERIES/FLOW//1HOUR/Ex1/")
+    update_hydrograph(
+        temp_file, "River: White  Reach: Muncie  RS: 15696.24", timeseries)
+    with h5py.File(temp_file, 'r') as temp:
+        dataset = temp["/Event Conditions/Unsteady/Boundary Conditions/Flow Hydrographs/River: White  Reach: Muncie  RS: 15696.24"]
+        assert dataset.shape == (7, 2)
+    os.remove(temp_file)
```
Comment (thwllms) [inline tests/test_hdf.py:73]: See above re: removing temporary test files

File: ras_remodeler.py [hunk 4]
```diff
@@ -1,39 +1,70 @@
...
+        timeseries = read_dss_timeseries(src_hydrograph)
+    elif input_type == 'CSV':
+        timeseries = read_csv_timeseries(src_hydrograph)
+    else:
+        raise ValueError(
+            "Invalid input_type option. Must be one of ['DSS', 'CSV']")
+    update_hydrograph(temp_hdf_filepath, plan_hdf_hydrograph_name,
+                      timeseries, keep_dates=keep_dates)
+    # overwrite existing file with new data
+    os.remove(plan_hdf)
```
Comment (thwllms) [inline ras_remodeler.py:65]: `os.remove` isn't going to work (I'm pretty sure) if `plan_hdf` points to `s3://` or `abfs://`. Although that's not a true requirement at this point. Probably leave as-is for this PR.

**Review by thwllms — COMMENTED** [review]
@barne856 looking good. A few things here and there.

We want to be sure that RAS is able to *run* with a `.p**.tmp.hdf` file that is produced with the `create-plan-tmp-hdf` command and has a hydrograph updated with `set-hdf-hydrograph` command.

I believe that can be tested like this (based on the `Compute Processes` section of a `.p**.hdf` file):
image

**Comment by barne856** [comment]
Added RasUnsteady binary and dependent libraries to repo with a test in test_set_plan_hdf_hydrograph() to run HEC-RAS with the .tmp.hdf file. This test runs in the dev container, but it's super slow. I'm pretty certain it works, but I didn't have the patience to wait for it to finish before pushing the code. I'll let run overnight and see if it completes...

**Comment by barne856** [comment]
I noticed that the example Muncie.p04.hdf has start and end dates for the hydrograph stored with the time set to 2400. This is a little confusing since I don't now if that is the end of the day or the beginning since 2400 and 0000 refer to the same time. Our string formatting will always produce 0000 instead of 2400 for that time, not sure if it matters or not for how HEC-RAS will interpret the data. Is this something that should be updated?

**Comment by barne856** [comment]
> Added RasUnsteady binary and dependent libraries to repo with a test in test_set_plan_hdf_hydrograph() to run HEC-RAS with the .tmp.hdf file. This test runs in the dev container, but it's super slow. I'm pretty certain it works, but I didn't have the patience to wait for it to finish before pushing the code. I'll let run overnight and see if it completes...

Came back after an hour and it did run successfully!

**Comment by thwllms** [comment]
@barne856 got this error:

```
vscode ➜ /workspaces/wat-ras-remodeler (brendan-dev) $ pytest
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.14, pytest-7.1.3, pluggy-1.0.0
rootdir: /workspaces/wat-ras-remodeler
collected 9 items                                                                                                                                                                            

tests/test_dss.py ..                                                                                                                                                                   [ 22%]
tests/test_hdf.py .....                                                                                                                                                                [ 77%]
tests/test_ras_remodeler.py .F                                                                                                                                                         [100%]

========================================================================================== FAILURES ==========================================================================================
________________________________________________________________________________ test_set_plan_hdf_hydrograph ________________________________________________________________________________

>   ???
E   AssertionError: assert 32512 == 0
E    +  where 32512 = <built-in function system>('export LD_LIBRARY_PATH=./bin/libs:./bin/libs/mkl && ./bin/RAS610/RasUnsteady ./tests/data/Muncie.c04 b04')
E    +    where <built-in function system> = os.system

/home/thomaswilliams3/dev/wat-ras-remodeler/tests/test_ras_remodeler.py:45: AssertionError
------------------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------------------
./bin/RAS610/RasUnsteady: error while loading shared libraries: libiomp5.so: cannot open shared object file: No such file or directory
================================================================================== short test summary info ===================================================================================
FAILED tests/test_ras_remodeler.py::test_set_plan_hdf_hydrograph - AssertionError: assert 32512 == 0
================================================================================ 1 failed, 8 passed in 0.88s =================================================================================
14:49:10.970      -----DSS---zopen   Existing file opened,  File: /tmp/3293400a-971b-4a0a-86f5-7acef90436c2.dss
14:49:10.970                         Handle 11;  Process: 25907;  DSS Versions - Software: 7-IE, File:  7-HL
14:49:10.970                         Single-user advisory access mode
14:49:10.970 -----DSS--- zread   Handle 11;  Version 1:  /REGULAR/TIMESERIES/FLOW/01Jul2019/1Hour/Ex1/
14:49:10.972      -----DSS---zclose  Handle 11;  Process: 25907;  File: /tmp/3293400a-971b-4a0a-86f5-7acef90436c2.dss
14:49:10.972                         Number records:         5
14:49:10.972                         File size:              15120  64-bit words
14:49:10.972                         File size:              118 Kb;  0 Mb
14:49:10.972                         Dead space:             3544
14:49:10.972                         Hash range:             4096
14:49:10.972                         Number hash used:       16
14:49:10.972                         Max paths for hash:     1
14:49:10.972                         Corresponding hash:     3976
14:49:10.972                         Number non unique hash: 0
14:49:10.972                         Number bins used:       16
14:49:10.972                         Number overflow bins:   0
14:49:10.972                         Number physical reads:  44
14:49:10.972                         Number physical writes: 0
14:49:10.972                         Number denied locks:    0
14:49:11.087      -----DSS---zopen   Existing file opened,  File: /tmp/d4f34a53-2b0a-48e4-b599-292c1637b7fa.dss
14:49:11.087                         Handle 11;  Process: 25907;  DSS Versions - Software: 7-IE, File:  7-HL
14:49:11.087                         Single-user advisory access mode
14:49:11.087 -----DSS--- zread   Handle 11;  Version 1:  /REGULAR/TIMESERIES/FLOW/01Jul2019/1Hour/Ex1/
14:49:11.088      -----DSS---zclose  Handle 11;  Process: 25907;  File: /tmp/d4f34a53-2b0a-48e4-b599-292c1637b7fa.dss
14:49:11.088                         Number records:         5
14:49:11.088                         File size:              15120  64-bit words
14:49:11.088                         File size:              118 Kb;  0 Mb
14:49:11.088                         Dead space:             3544
14:49:11.088                         Hash range:             4096
14:49:11.088                         Number hash used:       16
14:49:11.088                         Max paths for hash:     1
14:49:11.088                         Corresponding hash:     3976
14:49:11.088                         Number non unique hash: 0
14:49:11.088                         Number bins used:       16
14:49:11.088                         Number overflow bins:   0
14:49:11.088                         Number physical reads:  44
14:49:11.088                         Number physical writes: 0
14:49:11.088                         Number denied locks:    0
```
